golang/cosmos/x/swingset/keeper/iterator_helpers.go

```diff
@@ -0,0 +1,375 @@
+// both as byte slices.
+//
+// However to facilitate some usages, unlike an Iterator produced by a KVStore,
+// these iterators may yield a nil Value. This is useful for handling cases
+// like a stream of updates with some deletion operations.
+//
+// We define here the following:
+// - A generic implementation of an sdk.Iterator. It uses an underlying
+//   ExportDataReader which yields [2]string{key,value}, or [1]string{key}. For
+//   the latter, the Value of the iterator will be nil.
+//   The iterator implementation implements the sdk.Iterator protocol, and
+//   converts the key string to a byte slices for the Key, and the value string
+//   to a byte slice if present or nil otherwise for the Value
+// - 2 implementations of the ExportDataReader:
+//   - A reader which parses a jsonl stream from a ReadCloser (like a file)
+//   - A reader which uses a random access key/value source
+//     A source is a generic interface allowing reading string tuples at any
+//     position, with the reader doing bounds checks and keeping track of the
+//     current position.
+// - A helper to consume an iterator and write the "export data" into a Writer
+//   as line terminated json (jsonl-like compatible with both golang and JS)
+
+// ExportData is a [2]string{key,value} pair or [1]string{key} tuple
+// representing the key/value or deleted key of "export data" exchanged between
+// the iterator implementation and its underlying ExportDataReader
+// TODO: consider combining with vstorage StoreEntry
+type ExportData = []string
+
+// ExportDataReader is an abstraction for yielding key/value pairs as
+// strings. Used by exportDataIterator, the generic sdk.Iterator implementation.
+type ExportDataReader interface {
+	// Read returns the ExportData, or an error
+	Read() (ExportData, error)
+	// Close frees the underlying resource (usually a slice of some structure)
+	Close() error
+}
+
+// SafeIsIteratorValid is an helper to check whether an iterator is valid
+// while making sure any panic is reported as an error instead
+func SafeIsIteratorValid(iter sdk.Iterator) (valid bool, err error) {
+	defer helpers.RecoverToError(&err)
+	return iter.Valid(), nil
+}
+
+var _ sdk.Iterator = &exportDataIterator{}
+
+// exportDataIterator is the generic implementation of an sdk.Iterator which
+// leverages an ExportDataReader to access export data entries.
+type exportDataIterator struct {
+	reader  ExportDataReader
+	current ExportData
+	err     error
+}
+
+// NewExportDataIterator creates an iterator over a key/value string pair
+// reader. The values are iterated in order they're present in the source.
+func NewExportDataIterator(reader ExportDataReader) sdk.Iterator {
+	iter := &exportDataIterator{
+		reader: reader,
+	}
+	iter.Next()
+	return iter
+}
+
+// Domain implements Iterator
+func (iter *exportDataIterator) Domain() (start []byte, end []byte) {
+	return nil, nil
+}
+
+// Valid returns whether the current iterator is valid. Once invalid, the
```

Comment:
```suggestion
// Valid returns whether another pair can be requested from an iterator. Once invalid, the
```

Reply:
That is not correct. `Valid()` returns whether the iterator `.Value()` and `.Key()` can be called. Aka if the current position is valid.

Reply:
Ah, right, that's the weirdness of [their interface](https://pkg.go.dev/github.com/cosmos/cosmos-db#Iterator). But my goal here is to clarify for us what "valid" means.
```suggestion
// Valid returns whether the iterator has terminated (in which case Key() and
// Value() can no longer return results). Once invalid, the
```